lib/medtronic/medtronicSimulator.js

```diff
@@ -65,7 +83,220 @@ exports.make = function(config){
+            }else{
+             currBasal.payload = {duration : oldDuration};
+            }
+            if(currBasal.deliveryType === 'scheduled') {
+              currBasal.with_scheduleName(currPumpSettings.activeSchedule);
+            }
+            currBasal = currBasal.done();
+            events.push(currBasal);
+
+            var newSuppressed = {
+              type: 'basal',
+              deliveryType: 'scheduled',
+              rate: currSchedule[i].rate,
+              scheduleName : currPumpSettings.activeSchedule
+            };
+
+            var newBasal;
+            if(currBasal.deliveryType === 'temp') {
+              newBasal = config.builder.makeTempBasal();
+              if(currBasal.percent) {
+                newBasal.with_rate(currSchedule[i].rate * currBasal.percent)
+                       .with_percent(currBasal.percent);
+              } else {
+                newBasal.with_rate(currBasal.rate);
+              }
+            } else {
+              newBasal = config.builder.makeSuspendBasal();
+            }
+
+            var newJsDate = Date.parse(currBasal.deviceTime) + adjustedDuration;
+            newBasal.with_time(new Date(Date.parse(currBasal.time) + adjustedDuration).toISOString())
+              .with_duration(oldDuration - adjustedDuration)
+              .with_deviceTime(sundial.formatDeviceTime(newJsDate))
+              .with_timezoneOffset(currBasal.timezoneOffset)
+              .with_conversionOffset(currBasal.conversionOffset)
+              .set('index', currBasal.index);
+
+            if(currBasal.suppressed.suppressed) {
+              newBasal.set('suppressed', _.clone(currBasal.suppressed));
+              newBasal.suppressed.suppressed = newSuppressed;
+            } else {
+              newBasal.set('suppressed', newSuppressed);
+            }
+
+            annotate.annotateEvent(newBasal, 'medtronic/basal/fabricated-from-schedule');
+            setCurrBasal(newBasal);
+            changed = true;
+          }
+        }
+      };
+
+      var startTime = common.computeMillisInCurrentDay(currBasal);
+
+      var endTime = startTime + currBasal.duration;
+
+      if (endTime >= TWENTY_FOUR_HOURS) {
+        //check before midnight
+        checkSchedules(startTime, TWENTY_FOUR_HOURS - startTime, 0);
+        //check after midnight
+        checkSchedules(0, endTime - TWENTY_FOUR_HOURS, TWENTY_FOUR_HOURS - startTime);
+      } else {
+        checkSchedules(startTime, endTime, 0);
+      }
+    }
+    return changed;
+  }
+
   return {
+    basal: function(event) {
+      ensureTimestamp(event);
+
+      if (currBasal != null) {
+
+        if(!currBasal.isAssigned('duration')) {
+          // calculate current basal's duration
+          var duration = Date.parse(event.time) - Date.parse(currBasal.time);
+          currBasal.duration = duration;
+        }
+
+        if (event.deliveryType === 'temp' && event.rate == null && event.duration === 0) {
```

Comment:
Oh wow, I had to look twice before I saw it. It's true what they say about being so deep in your own code that you don't notice the obvious anymore. And indeed, readability much easier after that tiny refactor.